src/Tribe/Extension.php

```diff
@@ -0,0 +1,493 @@
+	 *      }
+	 *
+	 *      @type array $plugin_data If the plugin file header is parsed, the
+	 *                               resulting data is stored in this.
+	 * }
+	 */
+	protected $_args = array();
+
+	/**
+	 * The various extension instances
+	 *
+	 * If you directly modify this inside of a child class then you're arguably
+	 * an evil person.
+	 *
+	 * @var array {
+	 *      Each instance of an extension that extends this class
+	 *
+	 *      @type object $child_class_name instance
+	 * }
+	 */
+	private static $_instances = array();
+
+	/**
+	 * Get singleton instance of child class
+	 *
+	 * @param string $child_class (optional) Name of child class.
+	 *
+	 * @return object|null The extension's instance, or nothing if it can't be instantiated
+	 */
+	public static function instance( $child_class = null ) {
+		// Defaults to the name of the class that called this instance.
+		$child_class = empty( $child_class ) ? self::get_called_class() : $child_class;
+
+		// @see self::get_called_class() for the reason why this could be empty
+		if ( empty( $child_class ) ) {
+			return null;
+		}
+
+		if ( ! isset( self::$_instances[ $child_class ] ) ) {
+			self::$_instances[ $child_class ] = new $child_class();
+		}
+
+		return self::$_instances[ $child_class ];
+	}
+
+	/**
+	 * Initializes the extension
+	 *
+	 * Waits until after the init hook has fired.
+	 */
+	final private function __construct() {
+		$this->construct();
+
+		// The init() action/hook.
+		$init_hook = $this->get_init_hook();
+
+		// Continue plugin run after $init_hook has fired.
+		if ( did_action( $init_hook ) > 0 ) {
+			$this->register();
+		} else {
+			add_action( $init_hook, array( $this, 'register' ) );
+		}
+	}
+
+	/**
+	 * Empty function typically overriden by child class
+	 */
+	protected function construct() {}
+
+	/**
+	 * Adds a Tribe Plugin to the list of plugins this extension depends upon.
+	 *
+	 * If this plugin is not present or does not exceed the specified version
+	 * init() will not run.
+	 *
+	 * @param string      $main_class      The Main class for this Tribe plugin.
+	 * @param string|null $minimum_version Minimum acceptable version of plugin.
+	 */
+	final protected function add_required_plugin( $main_class, $minimum_version = null ) {
+		$this->set( array( 'requires', $main_class ), $minimum_version );
+	}
+
+	/**
+	 * Set the extension's tec.com URL
+	 *
+	 * @param string $url URL to the extension's page.
+	 */
+	final protected function set_url( $url ) {
+		$this->set( 'url', $url );
+		$this->add_meta_link( __( 'View Details' ), $url );
+	}
+
+	/**
+	 * Checks if the extension has permission to run, if so runs init() in child class
+	 */
+	final public function register() {
+		$is_plugin_authorized = tribe_register_plugin(
+			$this->get_plugin_file(),
+			$this->get( 'class', get_class( $this ) ),
+			$this->get_version(),
+			$this->get( 'requires', array() )
+		);
+
+		if ( $is_plugin_authorized ) {
+			$this->init();
```

Comment:
The reason constuct is present is because it is truly optional. Some extensions may choose to not declare it. But if they don't declare init() they have no reason to exist.